pkg/gui/controllers/helpers/fixup_helper.go

```diff
@@ -0,0 +1,175 @@
+			// If the commit is not found, it's most likely because it's already
+			// merged, and more than 300 commits away. Check if the last known
+			// commit is already merged; if so, show the "already merged" error.
+			return self.c.ErrorMsg("The base commit for this change is already on master")
+		}
+		// If we get here, the current branch must have more then 300 commits. Unlikely...
+		return self.c.ErrorMsg("Base commit is not in current view")
+	}
+	if commit.Status == models.StatusMerged {
+		return self.c.ErrorMsg("The base commit for this change is already on master")
+	}
+
+	if !useIndex {
+		if err := self.c.Git().WorkingTree.StageAll(); err != nil {
+			return err
+		}
+		_ = self.c.Refresh(types.RefreshOptions{Mode: types.SYNC, Scope: []types.RefreshableView{types.FILES}})
+	}
+
+	self.c.Contexts().LocalCommits.SetSelectedLineIdx(index)
+	return self.c.PushContext(self.c.Contexts().LocalCommits)
+}
+
+func (self *FixupHelper) getDiff() (string, bool, error) {
+	args := []string{"-U0", "--ignore-submodules=all", "HEAD", "--"}
+
+	// Try staged changes first
+	useIndex := true
+	diff, err := self.c.Git().Diff.DiffIndexCmdObj(append([]string{"--cached"}, args...)...).RunWithOutput()
+
+	if err == nil && diff == "" {
+		useIndex = false
+		// If there are no staged changes, try unstaged changes
+		diff, err = self.c.Git().Diff.DiffIndexCmdObj(args...).RunWithOutput()
+	}
+
+	return diff, useIndex, err
+}
+
+func (self *FixupHelper) parseDiff(diff string) []*deletedLineInfo {
+	lines := strings.Split(strings.TrimSuffix(diff, "\n"), "\n")
+
+	deletedLineInfos := []*deletedLineInfo{}
+
+	hunkHeaderRegexp := regexp.MustCompile(`@@ -(\d+)(?:,\d+)? \+\d+(?:,\d+)? @@`)
+
+	var filename string
+	var currentLineInfo *deletedLineInfo
+	finishHunk := func() {
+		if currentLineInfo != nil && currentLineInfo.numLines > 0 {
+			deletedLineInfos = append(deletedLineInfos, currentLineInfo)
+		}
+	}
+	for _, line := range lines {
+		if strings.HasPrefix(line, "diff --git") {
+			finishHunk()
+			currentLineInfo = nil
+		} else if strings.HasPrefix(line, "--- ") {
+			// For some reason, the line ends with a tab character if the file
+			// name contains spaces
+			filename = strings.TrimRight(line[6:], "\t")
+		} else if strings.HasPrefix(line, "@@ ") {
+			finishHunk()
+			match := hunkHeaderRegexp.FindStringSubmatch(line)
+			startIdx := utils.MustConvertToInt(match[1])
+			currentLineInfo = &deletedLineInfo{filename, startIdx, 0}
+		} else if currentLineInfo != nil && line[0] == '-' {
+			currentLineInfo.numLines++
+		}
+	}
+	finishHunk()
+
+	return deletedLineInfos
+}
+
+func (self *FixupHelper) blameDeletedLines(deletedLineInfos []*deletedLineInfo) []string {
+	var wg sync.WaitGroup
+	shaChan := make(chan string)
+
+	for _, info := range deletedLineInfos {
+		wg.Add(1)
+		go func(info *deletedLineInfo) {
+			defer wg.Done()
+
+			blameOutput, err := self.c.Git().Blame.BlameLineRange(info.filename, "HEAD", info.startLineIdx, info.numLines)
```

Comment:
Let me know if I understand what's happening here: we get the deleted line infos, and some of them are spread over several lines, and those lines may be from different commits. So for each set of lines we do a git blame to find out all the relevant commits. We then combine all of those together and return that as a slice of hashes.